### PR TITLE
[TEST] e2e test for new collective page - user profile

### PR DESCRIPTION
**Pull request by maddhruv**
Completes - https://github.com/opencollective/opencollective/issues/2450

Test Suites Includes
- Contributions section
  - Shows contributions with date since and amount contributed
  - Can load more
  - Can filter by contribution type (admin, financial...etc)
- Transactions section
  - Can filter by expense/contributions
  - Show transactions with all info and links

### CODE REVIEW

**Comment by maddhruv** [comment]
@Betree @znarf

**Comment by maddhruv** [comment]
Hey @Betree the tests pass in the local
don't know if the content is different in the environment the tests are running in CI

I have a question
Do I check for the content of the various DOMs or just look for the existence of the dom?

File: components/BudgetItemsList.js [hunk 1]
```diff
@@ -126,7 +126,7 @@ const BudgetItemsList = ({ items, isInverted }) => {
                   <Avatar collective={fromCollective} radius={40} />
                 </Box>
                 <Flex flexDirection="column" justifyContent="space-between">
-                  <P color="black.900" fontWeight="600">
+                  <P className="description" color="black.900" fontWeight="600">
```
Comment (Betree) [inline components/BudgetItemsList.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/BudgetItemsList.js [hunk 2]
```diff
@@ -135,7 +135,9 @@ const BudgetItemsList = ({ items, isInverted }) => {
                         id="Transactions.byWithGiftCard"
                         defaultMessage="by {collectiveName} with {collectiveGiftCardName} {giftCard} on {date}"
                         values={{
-                          collectiveName: <StyledLink as={LinkCollective} collective={fromCollective} />,
+                          collectiveName: (
+                            <StyledLink className="collective-link" as={LinkCollective} collective={fromCollective} />
```
Comment (Betree) [inline components/BudgetItemsList.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/BudgetItemsList.js [hunk 3]
```diff
@@ -146,7 +148,9 @@ const BudgetItemsList = ({ items, isInverted }) => {
                         id="Transactions.by"
                         defaultMessage="by {collectiveName} on {date}"
                         values={{
-                          collectiveName: <StyledLink as={LinkCollective} collective={fromCollective} />,
+                          collectiveName: (
+                            <StyledLink className="collective-link" as={LinkCollective} collective={fromCollective} />
```
Comment (Betree) [inline components/BudgetItemsList.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/BudgetItemsList.js [hunk 4]
```diff
@@ -164,7 +168,7 @@ const BudgetItemsList = ({ items, isInverted }) => {
                     −
                   </Span>
                 )}
-                <Span fontWeight="bold" mr={1}>
+                <Span className="transaction-amount" fontWeight="bold" mr={1}>
```
Comment (Betree) [inline components/BudgetItemsList.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/StyledButton.js [hunk 5]
```diff
@@ -53,7 +53,7 @@ const StyledButtonContent = styled.button`
 
 const StyledButton = ({ loading, ...props }) =>
   !loading ? (
-    <StyledButtonContent {...props} />
+    <StyledButtonContent className="filters" {...props} />
```
Comment (Betree) [inline components/StyledButton.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/StyledMembershipCard.js [hunk 6]
```diff
@@ -30,7 +30,7 @@ const getBackground = collective => {
 const StyledMembershipCard = ({ membership, intl, ...props }) => {
   const { collective, since, stats, role } = membership;
   return (
-    <StyledCard width={250} height={360} position="relative" {...props}>
+    <StyledCard className="collectives" width={250} height={360} position="relative" {...props}>
```
Comment (Betree) [inline components/StyledMembershipCard.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/StyledMembershipCard.js [hunk 7]
```diff
@@ -60,7 +60,7 @@ const StyledMembershipCard = ({ membership, intl, ...props }) => {
                   defaultMessage="{contributorType} since"
                   values={{ contributorType: formatMemberRole(intl, role) }}
                 />
-                <Span display="block" fontSize="LeadParagraph" fontWeight="bold">
+                <Span className="date-since" display="block" fontSize="LeadParagraph" fontWeight="bold">
```
Comment (Betree) [inline components/StyledMembershipCard.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/StyledMembershipCard.js [hunk 8]
```diff
@@ -74,7 +74,7 @@ const StyledMembershipCard = ({ membership, intl, ...props }) => {
                     </Span>
                   )}
                 </FormattedMessage>
-                <Span display="block" fontSize="LeadParagraph" fontWeight="bold">
+                <Span className="amount-contributed" display="block" fontSize="LeadParagraph" fontWeight="bold">
```
Comment (Betree) [inline components/StyledMembershipCard.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/collective-page/index.js [hunk 9]
```diff
@@ -216,7 +217,11 @@ class CollectivePage extends Component {
             isAnimated={true}
             onSectionClick={this.onSectionClick}
             LinkComponent={({ section, label, className }) => (
-              <a href={`#section-${section}`} className={className} onClick={e => e.preventDefault()}>
+              <a
+                href={`#section-${section}`}
+                className={classNames(`${className} navbar-${section}`)}
```
Comment (Betree) [inline components/collective-page/index.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/collective-page/sections/Contributions.js [hunk 10]
```diff
@@ -254,7 +254,7 @@ class SectionContributions extends React.PureComponent {
         ) : (
           <React.Fragment>
             <ContainerSectionContent>
-              <SectionTitle textAlign="left" mb={4}>
+              <SectionTitle id="Contributions" textAlign="left" mb={4}>
```
Comment (Betree) [inline components/collective-page/sections/Contributions.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

File: components/collective-page/sections/Contributions.js [hunk 11]
```diff
@@ -281,7 +281,12 @@ class SectionContributions extends React.PureComponent {
               </Flex>
               {nbMemberships < sortedMemberships.length && (
                 <Flex mt={3} justifyContent="center">
-                  <StyledButton textTransform="capitalize" minWidth={170} onClick={this.showMoreMemberships}>
+                  <StyledButton
+                    id="loadMore"
```
Comment (Betree) [inline components/collective-page/sections/Contributions.js]: As per https://docs.opencollective.com/help/developers/testing-with-cypress

> **Use data-cy to target DOM elements**
> 
> Don’t target elements based on CSS attributes such as: id, class, tag Add data-* attributes to make it easy to target elements

**Comment by maddhruv** [comment]
Yeah will make a fix 👍

**Comment by maddhruv** [comment]
Hey @Betree 
If I not visit `/xdamman/v2` the already created collective with good no. of contributions
I will not be able to get the `load more` button

**Comment by Betree** [comment]
The right solution would be to create these contributions, ideally by adding a new command dedicated to that to cypress. But that may be too much complexity for the bounty size, so either:
- Don't fill the `Load more` test, we'll complete it later
- Use an existing collective like `xdamman`, but keep in mind that your local data has probably been modified compared to the original dump used when you setup the API.

> Do I check for the content of the various DOMs or just look for the existence of the dom?

The goal of these tests for us is to avoid regressions. To properly test that we need to check the content as the DOM can be there while the content inside of it is broken.

**Comment by maddhruv** [comment]
> Don't fill the Load more test, we'll complete it later

+1

> Use an existing collective like xdamman, but keep in mind that your local data has probably been modified compared to the original dump used when you setup the API.

That shall also work

**Comment by maddhruv** [comment]
> To properly test that we need to check the content as the DOM can be there while the content inside of it is broken.

but how do I ensure the DB dump on my local matches that of available in CI

sometimes it threw errors on content matching

Note: have recently setup/restored the dump

**Comment by maddhruv** [comment]
@Betree 
i have removed the `load more` test as of now

File: test/cypress/integration/05-user_v2.test.js [hunk 12]
```diff
@@ -1,26 +1,38 @@
...
   describe('Contributions section', () => {
-    it.skip('Shows contributions with date since and amount contributed', () => {
-      // TODO
+    before(() => {
+      cy.visit('/xdamman/v2');
+      cy.get('[data-cy=Contributions]');
+    });
+    it('Shows contributions with date since and amount contributed', () => {
+      cy.get('[data-cy=collectives]').should('have.length.greaterThan', 1);
+      cy.get('[data-cy=date-since]').contains('since');
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: As per https://github.com/opencollective/opencollective-frontend/pull/2668#issuecomment-536454524, we also need to check the content as the DOM can be there while the content inside of it is broken.

File: test/cypress/integration/05-user_v2.test.js [hunk 13]
```diff
@@ -1,26 +1,38 @@
...
+      cy.get('[data-cy=amount-contributed]');
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: As per https://github.com/opencollective/opencollective-frontend/pull/2668#issuecomment-536454524, we also need to check the content as the DOM can be there while the content inside of it is broken.

File: test/cypress/integration/05-user_v2.test.js [hunk 14]
```diff
@@ -1,26 +1,38 @@
...
 
   describe('Transactions section', () => {
-    it.skip('Can filter by expense/contributions', () => {
-      // TODO
+    before(() => {
+      cy.visit('/xdamman/v2');
+    });
+    it('Can filter by expense/contributions', () => {
+      cy.get('[data-cy=navbar-contributions]').contains('Contributions');
+      cy.get('[data-cy=navbar-transactions]').contains('Transactions');
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: No need to test the navbar, this is about the transactions section.

But we do need to test the actual feature by changing the filters to ensure that the items displayed are filtered properly.

File: test/cypress/integration/05-user_v2.test.js [hunk 15]
```diff
@@ -1,26 +1,38 @@
...
     });
 
-    it.skip('Show transactions with all info and links', () => {
-      // TODO Has link to collective
-      // TODO Has transaction description, date and amount
+    it('Show transactions with all info and links', () => {
+      cy.get('[data-cy=transaction-details]');
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: As per https://github.com/opencollective/opencollective-frontend/pull/2668#issuecomment-536454524, we also need to check the content as the DOM can be there while the content inside of it is broken.

File: test/cypress/integration/05-user_v2.test.js [hunk 16]
```diff
@@ -1,26 +1,38 @@
...
+      cy.get('[data-cy=transaction-description]');
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: As per https://github.com/opencollective/opencollective-frontend/pull/2668#issuecomment-536454524, we also need to check the content as the DOM can be there while the content inside of it is broken.

File: test/cypress/integration/05-user_v2.test.js [hunk 17]
```diff
@@ -1,26 +1,38 @@
...
+      cy.get('[data-cy=transaction-details] > a').should('have.attr', 'href');
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: As per https://github.com/opencollective/opencollective-frontend/pull/2668#issuecomment-536454524, we also need to check the content as the DOM can be there while the content inside of it is broken.

File: test/cypress/integration/05-user_v2.test.js [hunk 18]
```diff
@@ -1,26 +1,38 @@
...
 
-    it.skip('Can load more', () => {
+    it('Can load more', () => {
       // TODO
     });
 
-    it.skip('Can filter by contribution type (admin, financial...etc)', () => {
-      // TODO
+    it('Can filter by contribution type (admin, financial...etc)', () => {
+      cy.get('button[data-cy=filters]').contains(/(Core|Financial|Events)/);
```
Comment (Betree) [inline test/cypress/integration/05-user_v2.test.js]: We do need to test the actual feature by changing the filters to ensure that the items displayed are filtered properly.

**Comment by Betree** [comment]
> but how do I ensure the DB dump on my local matches that of available in CI
> 
> sometimes it threw errors on content matching
> 
> Note: have recently setup/restored the dump

CI starts with a fresh dump, but other cypress tests may affect the data. However because this one runs at the very beginning, it should be safe to assume that a fresh dump will match.

**Comment by Betree** [comment]
Also per my comment https://github.com/opencollective/opencollective-frontend/pull/2668#issuecomment-536454524

Please don't ignore the `Load more` test if you're using an existing collective (like `xdamman`). What I meant by this:

> so either:
> 
> - Don't fill the Load more test, we'll complete it later
> - Use an existing collective like xdamman, but keep in mind that your local data has probably been modified compared to the original dump used when you setup the API.
> 

Was that you could ignore this test if you chose the solution of `cy.createHostedCollective`. If you're going with an existing collective this is actually pretty simple to test, you just need to check that the button exists and that it loads more collectives when clicked.

**Comment by Betree** [comment]
@maddhruv any news on that one? Can I help with anything?

**Comment by maddhruv** [comment]
I will pick this up tonight, didn't get enough time to focus up on this

**Comment by maddhruv** [comment]
closing in favour of https://github.com/opencollective/opencollective-frontend/pull/2756
as the base `repo/branch` here doesn't exist anymore

**Pull request closed without merging**